keyserver/server.go

```diff
@@ -0,0 +1,183 @@
+	Address              string  `toml:"address"` // address:port
+	LoadedHistoryLength  uint64  `toml:"loaded_history_length"`
+	RegistrationCapacity uint64  `toml:"registration_capacity"`
+	PoliciesPath         string  `toml:"policies"`
+	TLS                  tlsInfo `toml:"tls"`
+}
+
+type tlsInfo struct {
+	TLSCertPath string `toml:"cert"`
+	TLSKeyPath  string `toml:"key"`
+}
+
+type ConiksServer struct {
+	sync.Mutex
+	directory *protocol.ConiksDirectory
+	tbs       map[string]*merkletree.TemporaryBinding
+
+	stop     chan struct{}
+	waitStop sync.WaitGroup
+
+	db        kv.DB
+	secretKey sign.PrivateKey
+
+	policies         merkletree.Policies
+	policiesFilePath string
+	policiesMutex    sync.Mutex
+	reloadChan       chan os.Signal
+	epochTimer       *time.Timer
+}
+
+func LoadServerConfig(path string) *ServerConfig {
+	var conf ServerConfig
+	if _, err := toml.DecodeFile(path, &conf); err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+		return nil
+	}
+	return &conf
+}
+
+func New(conf *ServerConfig) *ConiksServer {
+	// load signing key
+	skBytes, err := ioutil.ReadFile(conf.SigningKeyPath)
+	if err != nil {
+		log.Fatalf("Cannot read signing key: %v", err)
+		return nil
+	}
+	if len(skBytes) != sign.PrivateKeySize {
+		log.Fatalf("Signing key must be 64 bytes (got %d)", len(skBytes))
+		return nil
+	}
+
+	sk := make([]byte, sign.PrivateKeySize)
+	copy(sk, skBytes[:sign.PrivateKeySize])
+
+	// open db
+	kvdb := leveldbkv.OpenDB(conf.DatabasePath)
+
+	// read server policies
+	p, err := readPolicies(conf.PoliciesPath)
+	if err != nil {
+		log.Fatalf("Cannot read policies config: %v", err)
+		return nil
+	}
+	policies := merkletree.NewPolicies(p.EpochDeadline, p.VRFKey)
+
+	// create server instance
+	server := new(ConiksServer)
+	server.stop = make(chan struct{})
+	server.db = kvdb
+	server.secretKey = sk
+	server.policies = policies
+	server.policiesFilePath = conf.PoliciesPath
+	server.reloadChan = make(chan os.Signal, 1)
+	signal.Notify(server.reloadChan, syscall.SIGUSR2)
+	server.epochTimer = time.NewTimer(time.Duration(policies.EpDeadline()) * time.Second)
+	server.directory = protocol.InitDirectory(policies, sk, conf.LoadedHistoryLength)
+	server.tbs = make(map[string]*merkletree.TemporaryBinding, conf.RegistrationCapacity)
+
+	return server
+}
+
+func (server *ConiksServer) RunWithConfig(conf *ServerConfig) {
+	server.waitStop.Add(1)
+	go server.EpochUpdate()
+
+	// server listener
+	cer, err := tls.LoadX509KeyPair(conf.TLS.TLSCertPath, conf.TLS.TLSKeyPath)
+	if err != nil {
+		panic(err)
+	}
+
+	config := &tls.Config{Certificates: []tls.Certificate{cer}}
+	// server public port
+	ln, err := tls.Listen("tcp", conf.Address, config)
+	if err != nil {
+		panic(err)
+	}
+
+	// server regitration port
+	regLn, err := tls.Listen("tcp", conf.RegistrationAddress, config)
+	if err != nil {
+		panic(err)
+	}
+
+	server.waitStop.Add(1)
+	go server.listenForRequests(ln)
+
+	server.waitStop.Add(1)
+	go server.listenForRequests(regLn)
```

Comment:
Can you help me here a little: If I understand this correctly, it means that we start the same server on two different ports. Currently, the server handles all message with `ConiksServer.handleClientMessage`. On this branch it this can only handle messages of type `RegistrationRequest`. So basically a client can register connecting to one port or the other. Is that correct?